### PR TITLE
improve match bracket matching

### PR DESCRIPTION
**Pull request by pascalkuthe**
This PR fixes a couple of edgecases that were noticed in the match bracket implementation after the recent changes to match_bracket.

This should cover both the markdown cases discovered by @David-Else (although I think there are still some cases involving injections that may not be covered but all the examples discussed work now) and the python cases discoreved by @ thomasaarholt. I also included fixes for some issues for cases where in cases where multiple nested pairs are siblings. I think this probably shouldn't happen usually, but it can happen in case there are syntax error so it's nice to be a bit more robust

### CODE REVIEW

File: helix-core/src/match_brackets.rs [hunk 1]
```diff
@@ -274,27 +261,37 @@ fn as_close_pair(doc: RopeSlice, node: &Node) -> Option<char> {
...
+                }
+                depth -= 1;
+            } else if c == start_char {
+                depth += 1;
+            }
+            None
         })
 }
 
 /// Tests if this node is a pair close char and returns the expected open char
```
Comment (woojiq) [inline helix-core/src/match_brackets.rs:289]: Outdated function description.

File: helix-core/src/match_brackets.rs [hunk 2]
```diff
@@ -241,29 +245,12 @@ pub fn is_valid_pair(ch: char) -> bool {
...
-    let end_byte = node.end_byte().saturating_sub(1);
-
-    if start_byte >= len || end_byte >= len {
-        return None;
-    }
-
-    Some((start_byte, end_byte))
-}
-
 /// Tests if this node is a pair close char and returns the expected open char
```
Comment (woojiq) [inline helix-core/src/match_brackets.rs:248]: Outdated function description.

**Review by woojiq — COMMENTED** [review]
I will rebase my changes on top of yours to see if they work together.
Unfortunately I can't review your PR properly because I don't really know how to work with the tree care API, I just used `find_pair` in my commits)

**Comment by David-Else** [comment]
Brilliant! In Markdown words surrounded in backticks now work everywhere, and also all the individual matching pair elements (`[]` `()` `""`) inside a link work, for example:
```
My favorite search engine is [Duck Duck Go](https://duckduckgo.com "The best search engine for privacy").
```

For reasons unknown to me you can `mim` inside the `"` in the link above, but not in general text:

```
Try `mim in this backticked text`, no problem, but it doesn't work in "this quoted text"
```
Not that big a deal, but I thought I would mention it in case it was a bug.

**Comment by woojiq** [comment]
> For reasons unknown to me you can `mim` inside the `"` in the link above, but not in general text:

Type `:tree-sitter-subtree` and you will notice that there are no nodes for default text, but there is a node for "link_title" that contains quotes.

**Comment by pascalkuthe** [comment]
Markdown seems to be just a weird case its grammar the quotes there are not actually part of the sytnax tree (they are really just plaintext). The inline grammar that markdown uses is pretty hacky and the way it seems to produce nodes... Frankly doesn't make much sense. I think it works this way because of TS limitations/because markdown really isn't all that suited to a proper syntax tree but its mostly a coincidence that is works sometimes at all that is down to implementation details of the language.

`mim`/`mam`/`mm` for quotes (and even brackets where they aren't part of links) just won't work with markdown. The fact that they sometimes do is pure coincidence.

**Comment by pascalkuthe** [comment]
for markdown in particularl we may consider somekind of switch so that the plaintext version of `mam` is used in the longterm but that would require new language specific config so that's probably something left to the more distant future

**Review by thomasaarholt — APPROVED** [review]
Tested this with #10611 on top, only with Python. Everything works as expected. Nice work, and thanks for fixing!

**Comment by archseer** [comment]
Can we add some unit/integration tests for the bugs being fixed?

**Comment by pascalkuthe** [comment]
good point I added a couple integration tests

File: helix-term/tests/test/commands/movement.rs [hunk 3]
```diff
@@ -806,3 +806,129 @@ async fn test_select_prev_sibling() -> anyhow::Result<()> {
...
+                fn foo() {
+                    foo::<bar<>#[>|]# 
+                }
+            "##},
+        ),
+    ];
+
+    let python_tests = vec![
+        // python quotes have a slightly more complex syntax tree
+        // that tirggerd a bug in an old implementation so we test
```
Comment (the-mikedavis) [inline helix-term/tests/test/commands/movement.rs:902]:
```suggestion
        // that triggerd a bug in an old implementation so we test
```

**Review by archseer — APPROVED** [review]

**Review by the-mikedavis — APPROVED** [review]

**Pull request merged**